## Problema Cazatalentos buscando a “La verdadera mejor encestadora”

In [ ]:
import  numpy as np
import scipy.stats as stats
import math

In [ ]:
np.random.seed(350003)

# calcula cuantos encestes logra un jugadora con indice de enceste prob
# haciendo qyt tiros libres

def ftirar(prob, qty):
  return sum(np.random.rand(qty) < prob)

### Cazatalentos 1

In [ ]:
# defino los jugadoras
mejor = 0.7 # Se considera a la mejor jugadora con un porcentaje de aciertos de 0.7
peloton = np.array(range(501, 600)) / 1000 # Genero las 99 jugadoras restantes asumiendo que su porcentaje de aciertos va desde 0.51 a 0.6
jugadoras = np.append(mejor, peloton)

# vectorizo la funcion  ftirar

vec_ftirar = np.vectorize(ftirar)

primera_ganadora = 0

for i in range(1000): # hago mil experimentos
  vaciertos = vec_ftirar(jugadoras, 100) # 100 tiros libres cada jugadora
  mejor_ronda = np.argmax(vaciertos)
  if mejor_ronda == 0:
    primera_ganadora += 1

print("Cantidad de veces que gano la mejor jugadora: ",(primera_ganadora) )
print("Porcentaje de efectividad de este metodo ", primera_ganadora/1000 )


Cantidad de veces que gano la mejor jugadora:  611
Porcentaje de efectividad de este metodo  0.611


Según el calculo realizado, la probabilidad de hacer tirar 100 tiros a 100 jugadoras y que la mejor jugadora gane es de 0.611

Orden: C1

###Cazatalentos 2

In [ ]:
# defino los jugadoras
mejor_2= 0.7
# En este caso son 200 jugadoras las que participan de la prueba por lo que aumento la cantidad de valores en la variable peloton
peloton_2= np.array(range(421, 620)) / 1000 # Dado que participan mas jugadoras asumo que el porcentaje de acierto de estas aumenta del rango (0.5 - 0.6) a (0.42 - 0.62)
jugadoras_2 = np.append(mejor_2, peloton_2)


# vectorizo la funcion  ftirar

vec_ftirar_2 = np.vectorize(ftirar)

primera_ganadora_2 = 0

for i in range(1000): #  mil experimentos
  vaciertos_2 = vec_ftirar_2(jugadoras_2, 100) # 100 tiros libres cada jugadora
  mejor_ronda_2 = np.argmax(vaciertos_2)
  if mejor_ronda_2 == 0:
    primera_ganadora_2 += 1

print("Cantidad de veces que gano la mejor jugadora: ",(primera_ganadora_2) )
print("Porcentaje de efectividad de este metodo ", primera_ganadora_2/1000 )

Cantidad de veces que gano la mejor jugadora:  468
Porcentaje de efectividad de este metodo  0.468


Para la cazatalentos 2 la probabilidad de encontrar a la mejora jugadora baja con respecto al caso de la cazatalentos 1, esto se debe a que al aumentar la cantidad de jugadoras que participan de la prueba aumenta la probabilidad de que una de ellas tenga suerte en la prueba y supere a la mejor jugadora. Esta baja en la efectividad tambien se debe al aumento en el porcentaje de enceste del peloton de 199 de jugadoras.

Orden: C2 < C1

### Cazatalentos 3

In [ ]:
# defino las jugadoras
# Manteniendo los valores de pruebas previas asumo que la mejor jugadora tienen una probabilidad de acierto de 0.7 y la otra de 0.6
jugadora_1 = 0.7
jugadora_2 = 0.6
ambas_jugadoras = np.append(jugadora_1, jugadora_2)

# vectorizo la funcion  ftirar

vec_ftirar_3 = np.vectorize(ftirar)

primera_ganadora_3 = 0

for i in range(1000): #  mil experimentos
  vaciertos_3 = vec_ftirar_3(ambas_jugadoras, 100) # 100 tiros libres cada jugadora
  mejor_ronda_3 = np.argmax(vaciertos_3)
  if mejor_ronda_3 == 0:
    primera_ganadora_3 += 1

print("Cantidad de veces que gano la mejor jugadora: ",(primera_ganadora_3) )
print("Porcentaje de efectividad de este metodo ", primera_ganadora_3/1000 )

Cantidad de veces que gano la mejor jugadora:  940
Porcentaje de efectividad de este metodo  0.94


En este caso al ser solo 2 jugadoras aumenta la probabilidad de que la jugadora con mayor porcentaje de acierto gane la prueba, dado que solo tiene que superar a 1 jugadora y la probabilidad de que una jugadora del peloton tenga mas suerte y le gane disminuye por que es solo una.
Por lo tanto hay 94% de probabilidad que tenga 0.7 de acierto

Orden: C2 < C1 < C3

### Cazatalentos 4

In [ ]:
# defino las jugadoras
mejor_4= 0.7
peloton_4= np.array(range(501, 600)) / 1000 # son 100 jugadoras
jugadoras_4 = np.append(mejor_4, peloton_4)

# vectorizo la funcion  ftirar

vec_ftirar_4 = np.vectorize(ftirar)

primera_ganadora_4 = 0

for i in range(1000): # mil experimentos
  vaciertos_4 = vec_ftirar_4(jugadoras_4, 10) # 10 tiros libres cada jugadora
  mejor_ronda_4 = np.argmax(vaciertos_4)
  if mejor_ronda_4 == 0:
    primera_ganadora_4 += 1

print("Cantidad de veces que gano la mejor jugadora: ",(primera_ganadora_4) )
print("Porcentaje de efectividad de este metodo ", primera_ganadora_4/1000 )

Cantidad de veces que gano la mejor jugadora:  163
Porcentaje de efectividad de este metodo  0.163


En este caso al disminuir la cantidad de tiros libres que tiene que tirar cada jugadora aumenta la probabilidad de que una de las 99 jugadoras dentro del peloton tenga suerte y obtenga una mayor puntuación que la mejor jugadora

Orden: C4 < C2 < C1 < C3

### Cazatalentos 5

Al hacer un primer corte donde solo se queda con las 5 mejores jugadoras, se debe revisar la probabilidad de que la mejor jugadora haya quedado dentro de esas 5 y avanzado a las rondas 2 y 3

In [ ]:
# Defino los Parámetros necesarios
num_simulaciones = 1000  # Número de simulaciones
num_jugadoras = 99  # Número de otras jugadoras
num_tiros = 100  # Número de tiros
porcentaje_mejor = 0.70  # Porcentaje de acierto de la mejor jugadora
porcentaje_otras = np.random.uniform(0.5, 0.6, num_jugadoras)  # Porcentaje de acierto de las otras jugadoras
num_mejores = 5  # Número de mejores jugadoras que deseas contar

# Función para simular una ronda de tiros y contar cuántas veces la mejor jugadora está entre las mejores 5 resultados
def simular_ronda():
    tiros_mejor = np.random.rand(num_tiros) < porcentaje_mejor
    tiros_otras = np.random.rand(num_jugadoras, num_tiros) < porcentaje_otras[:, np.newaxis]
    aciertos_mejor = np.sum(tiros_mejor)
    aciertos_otras = np.sum(tiros_otras, axis=1)
    mejores = np.argsort(aciertos_otras)[-num_mejores:]
    return aciertos_mejor >= np.min(aciertos_otras[mejores])

# Realizar las mil simulaciones
exitos = sum([simular_ronda() for _ in range(num_simulaciones)])
probabilidad_5 = exitos / num_simulaciones

print("Probabilidad de que la mejor tiradora esté entre las 5 mejores:", probabilidad_5)

Probabilidad de que la mejor tiradora esté entre las 5 mejores: 0.908


In [ ]:
# Ahora de las 5 quiero saber la probabilidad de elegir la mejor entre esas 5
# defino los jugadoras
mejor = 0.7 # Se considera a la mejor jugadora con un porcentaje de aciertos de 0.7
peloton = [0.6, 0.6, 0.6, 0.6] # dado que pasaron el corte asumo que tienen el porcentaje maximo de acierto dentro del peloton
jugadoras = np.append(mejor, peloton)

# vectorizo la funcion  ftirar

vec_ftirar = np.vectorize(ftirar)

primera_ganadora = 0

for i in range(1000): # hago mil experimentos
  vaciertos = vec_ftirar(jugadoras, 300) # 300 tiros libres en tota cada jugadora
  mejor_ronda = np.argmax(vaciertos)
  if mejor_ronda == 0:
    primera_ganadora += 1

print("Cantidad de veces que gano la mejor jugadora: ",(primera_ganadora) )
probabilidad_ganadora = primera_ganadora / 1000
print("Porcentaje de efectividad ", probabilidad_ganadora)


Cantidad de veces que gano la mejor jugadora:  988
Porcentaje de efectividad  0.988


In [ ]:
# Para obtener la probabilida de este metodo debo multiplicar las 2 probabilidades obtenidas, es decir la probabilidad de que pase el primer corte por la probabilidad de que supere a las otras 4 jugadoras

print("porcentaje de efectividad total del metodo: ", probabilidad_5 * probabilidad_ganadora)

porcentaje de efectividad total del metodo:  0.897104


Orden: C4 < C2 < C1 < C5 < C3


### Cazatalentos 6

In [ ]:


"""
El caso del cazatalentos 6 se podria pensar de forma bayesiana sabiendo que la jugadora elegida logro encestar 80 de 100 tiros
cual es la probabilidad que sea  la jugadora con porcentaje de acierto 0.7, es decir, la mejor jugadora y no una de las 99 jugadoras con
porcentaje de acierto de 0.5 a 0.6
"""
# Generar 99 valores de probabilidad aleatorios en el rango [0.5, 0.6]
probabilidades_peloton = np.random.uniform(0.5, 0.6, 99)

# Probabilidad de elegir a la jugadora con 0.7 de acierto
P_mejor= 1 / 100

# Probabilidad de que la jugadora acierte 80 tiros si tiene un 0.7 de acierto
P_80_dado_mejor = stats.binom.pmf(80, 100, 0.7)
print(P_80_dado_mejor)

# Probabilidad de que la jugadora acierte 80 tiros si no tiene un 0.7 de acierto
P_80_dado_peloton = 0
for p in probabilidades_peloton:
    P_80_dado_peloton += stats.binom.pmf(80, 100, p)

print(P_80_dado_peloton)

# Probabilidad de no elegir a la jugadora con 0.7 de acierto
P_peloton = 99 / 100

# Con el teorema de Bayes se calcula la probabilidad de que sea la mejor jugadora dado que encesto 80 tiros
P_mejor_dado_80 = (P_80_dado_mejor * P_mejor) / ((P_80_dado_mejor * P_mejor) + (P_80_dado_peloton * P_peloton))

print("La probabilidad de que la jugadora elegida tenga 0.7 de acierto dado que acertó 80 tiros es {}".format((P_mejor_dado_80)))

0.007575644925726648
0.00011344278072628454
La probabilidad de que la jugadora elegida tenga 0.7 de acierto dado que acertó 80 tiros es 0.40282098713109493


Orden: C4 < C6 < C2 < C1 < C5 < C3

### Cazatalentos 7

In [ ]:
# Parámetros
num_simulaciones = 1000  # Número de simulaciones
num_jugadoras = 99  # Número de otras jugadoras
num_tiros = 100  # Número de tiros
porcentaje_mejor = 0.70  # Porcentaje de acierto de la mejor jugadora
porcentaje_otras = np.random.uniform(0.5, 0.6, num_jugadoras)  # Porcentaje de acierto de las otras jugadoras
num_mejores = 5  # quiero evaluar si queda entre las 5 mejores

# Función para simular una ronda de tiros y contar cuántas veces la mejor jugadora está entre las mejores
def simular_ronda():
    tiros_mejor = np.random.rand(num_tiros) < porcentaje_mejor
    tiros_otras = np.random.rand(num_jugadoras, num_tiros) < porcentaje_otras[:, np.newaxis]
    aciertos_mejor = np.sum(tiros_mejor)
    aciertos_otras = np.sum(tiros_otras, axis=1)
    mejores = np.argsort(aciertos_otras)[-num_mejores:]
    return aciertos_mejor >= np.min(aciertos_otras[mejores])

# Hago las 1000 simulaciones y me quedo con la probabilidad de que quede entre las 5 mejores, esto es lo mismo que en el cazatalentos 5
exitos = sum([simular_ronda() for _ in range(num_simulaciones)])
probabilidad_5_mejores = exitos / num_simulaciones

print("Probabilidad de que la mejor tiradora esté entre las 5 mejores:", probabilidad_5_mejores)

# Parámetros
num_simulaciones = 1000  # Número de simulaciones
num_jugadoras = 5  # Número de jugadoras
num_tiros = 100  # Número de tiros

# Función para simular una ronda de tiros y contar cuántas veces la mejor jugadora encesta más
def simular_ronda():
    porcentajes_acierto = np.array([0.7, 0.6, 0.6, 0.6, 0.6])  # Porcentajes de acierto de las jugadoras, supongo 0.6 de porcentaje de acierto para las otras 4 jugadoras
    tiros = np.random.rand(num_jugadoras, num_tiros) < porcentajes_acierto[:, np.newaxis]
    aciertos = np.sum(tiros, axis=1)
    return aciertos.argmax() == 0

# Hago las 1000 simulaciones y evaluo la probabilidad de que la mejor jugadora sea la que obtenga el mejor resultado en esta segunda tanda de tiros
exitos = sum([simular_ronda() for _ in range(num_simulaciones)])
probabilidad_la_mejor = exitos / num_simulaciones

print("Probabilidad de que la jugadora con el porcentaje de acierto más alto sea la que más encesta:", probabilidad_la_mejor)

print("La probabilidad total del metodo sería la multiplicación entre ambas probabilidades: ", probabilidad_5_mejores * probabilidad_la_mejor )

Probabilidad de que la mejor tiradora esté entre las 5 mejores: 0.909
Probabilidad de que la jugadora con el porcentaje de acierto más alto sea la que más encesta: 0.825
La probabilidad total del metodo sería la multiplicación entre ambas probabilidades:  0.749925


Orden: C4 < C6 < C2 < C1 < C7 < C5 < C3

### Cazatalentos 8

Dado que en el enunciado se informa que la jugadora seleccionada tenia una probabilidad de acierto de 0.79 tomo eso como indicador del metodo de seleccion para comparar con los otros cazatalentos

Orden: C4 < C6 < C2 < C1 < C8 < C5 < C3 < C8

### Cazatalentos 9

In [ ]:
# Dado que tengo la información de todos los modelos o todas las rondas de tiro de la jugadora hago un promedio del porcentaje de encestes entre todas las rondas para obtener un estimador mas real
# Valores de encestes en cada ronda
aciertos = np.array([68, 74, 78, 70, 68, 63, 80, 68, 67, 65])

# Calcular la media de encestes
media_aciertos= np.mean(aciertos)

print("Media del porcentaje de acierto segun las 10 rondas ejecutadas:", media_aciertos/100)



Media del porcentaje de acierto segun las 10 rondas ejecutadas: 0.701


Basandome en la probabilidad de acierto obtenida en cada metodo puedo establecer una orden para las jugadoras seleccionadas:

Orden: C4 < C6 < C2 < C1 < C9 < C7  < C5 < C3 < C8